File: agent-python/src/pi/tools/read.py
# src/pi/tools/read.py
from __future__ import annotations

import asyncio
from pathlib import Path
from typing import Any, Callable

from .base import Tool, ToolResult

MAX_LINES = 2000
MAX_BYTES = 200 * 1024  # 200KB


class ReadTool(Tool):
    @property
    def name(self) -> str:
        return "read"

    @property
    def description(self) -> str:
        return (
            "Read the contents of a file. Returns line-numbered content. "
            "Use offset and limit for large files."
        )

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "path": {"type": "string", "description": "Path to file (relative or absolute)"},
                "offset": {"type": "number", "description": "Line number to start from (1-indexed)"},
                "limit": {"type": "number", "description": "Maximum number of lines to read"},
            },
            "required": ["path"],
        }

    async def execute(
        self,
        tool_call_id: str,
        args: dict[str, Any],
        signal: asyncio.Event | None = None,
        on_update: Callable[[str], None] | None = None,
    ) -> ToolResult:
        path_str = args.get("path", "")
        offset = int(args.get("offset", 1))
        limit = args.get("limit")

        path = Path(path_str)
        if not path.exists():
            return ToolResult.err(f"File not found: {path_str}")
        if not path.is_file():
            return ToolResult.err(f"Not a file: {path_str}")

        try:
            raw = path.read_bytes()
        except OSError as e:
            return ToolResult.err(f"Cannot read {path_str}: {e}")

        byte_truncated = len(raw) > MAX_BYTES
        if byte_truncated:
            raw = raw[:MAX_BYTES]

        text = raw.decode("utf-8", errors="replace")
        all_lines = text.splitlines()
        total = len(all_lines)

        start = max(0, offset - 1)
        end = start + (int(limit) if limit is not None else MAX_LINES)
        end = min(end, start + MAX_LINES)
        slice_lines = all_lines[start:end]

        lines_shown_start = start + 1
        lines_shown_end = start + len(slice_lines)

        numbered = "\n".join(f"{lines_shown_start + i}\t{line}" for i, line in enumerate(slice_lines))

        truncated = end < total or byte_truncated
        if truncated:
            numbered += f"\n[Showing lines {lines_shown_start}-{lines_shown_end} of {total}. Use offset={lines_shown_end + 1} to continue.]"

        return ToolResult.ok(numbered)

File: agent-rust/src/tools/read.rs
use async_trait::async_trait;
use tokio_util::sync::CancellationToken;
use crate::tools::{Tool, ToolResult, UpdateFn};

pub struct ReadTool;

const MAX_LINES: usize = 2000;
const MAX_BYTES: usize = 200 * 1024;

#[async_trait]
impl Tool for ReadTool {
    fn name(&self) -> &str { "read" }
    fn description(&self) -> &str {
        "Read file contents with 1-indexed line numbers. \
         offset=start line (1-indexed), limit=max lines. \
         Returns: 'N\\tcontent\\n' per line."
    }
    fn schema(&self) -> serde_json::Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "path":   {"type": "string"},
                "offset": {"type": "number", "description": "Start line (1-indexed, default 1)"},
                "limit":  {"type": "number", "description": "Max lines (default 2000)"}
            },
            "required": ["path"]
        })
    }

    async fn execute(
        &self,
        _id: &str,
        args: serde_json::Value,
        _cancel: CancellationToken,
        _on_update: Option<&UpdateFn>,
    ) -> ToolResult {
        let path = match args["path"].as_str() {
            Some(p) => p.to_string(),
            None => return ToolResult::err("missing 'path' argument"),
        };
        let offset = args["offset"].as_u64().unwrap_or(1).max(1) as usize;
        let limit  = args["limit"].as_u64().map(|n| n as usize).unwrap_or(MAX_LINES);

        let content = match tokio::fs::read_to_string(&path).await {
            Ok(c)  => c,
            Err(e) => return ToolResult::err(format!("cannot read {path}: {e}")),
        };

        let all_lines: Vec<&str> = content.lines().collect();
        let total = all_lines.len();

        // offset is 1-indexed; convert to 0-indexed start
        let start = (offset - 1).min(total);
        let end   = (start + limit).min(total).min(start + MAX_LINES);

        let mut output = String::new();
        let mut bytes  = 0usize;
        let mut actual_end = start;

        for (i, line) in all_lines[start..end].iter().enumerate() {
            let formatted = format!("{}\t{}\n", start + i + 1, line);
            bytes += formatted.len();
            if bytes > MAX_BYTES {
                output.push_str("[...truncated at 200KB limit...]\n");
                break;
            }
            output.push_str(&formatted);
            actual_end = start + i + 1;
        }

        if actual_end < total {
            output.push_str(&format!(
                "[Showing lines {}-{} of {}. Use offset={} to continue.]\n",
                start + 1, actual_end, total, actual_end + 1
            ));
        }

        ToolResult::ok(output)
    }
}
